core/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java

```diff
@@ -139,19 +145,30 @@ public void prepareSelectedBuckets(long... selectedBuckets) throws IOException {
         this.selectedBuckets = hash;
 
         collector.preCollection();
-        if (collector.needsScores()) {
-            throw new IllegalStateException("Cannot defer if scores are needed");
+        boolean needsScores = collector.needsScores();
+        Weight weight = null;
+        if (needsScores) {
+            weight = aggContext.searchContext().searcher()
+                        .createNormalizedWeight(aggContext.searchContext().query(), true);
         }
-
         for (Entry entry : entries) {
             final LeafBucketCollector leafCollector = collector.getLeafCollector(entry.context);
-            leafCollector.setScorer(Lucene.illegalScorer("A limitation of the " + SubAggCollectionMode.BREADTH_FIRST
-                    + " collection mode is that scores cannot be buffered along with document IDs"));
+            DocIdSetIterator docIt = null;
+            if (needsScores && entry.docDeltas.size() > 0) {
+                Scorer scorer = weight.scorer(entry.context);
+                docIt = scorer.iterator();
+                leafCollector.setScorer(scorer);
+            } else {
+                leafCollector.setScorer(null);
```

Comment:
Sure though if needsScores is false then it should guarantee that the sub aggregations will never access the scorer ?